File: src/main/java/org/hildan/minecraft/mining/optimizer/ore/WorldGenMinable.java
package org.hildan.minecraft.mining.optimizer.ore;

import org.hildan.minecraft.mining.optimizer.blocks.BlockType;
import org.hildan.minecraft.mining.optimizer.blocks.Sample;

import java.util.Random;

@SuppressWarnings({"MagicNumber", "UnnecessaryExplicitNumericCast"})
class WorldGenMinable {

    private final int maxVeinSize;

    private final BlockType blockTypeToGenerate;

    private final BlockType blockTypeToReplace;

    public WorldGenMinable(BlockType blockTypeToGenerate) {
        this.maxVeinSize = blockTypeToGenerate.getMaxVeinSize();
        this.blockTypeToGenerate = blockTypeToGenerate;
        this.blockTypeToReplace = BlockType.STONE;
    }

    @SuppressWarnings({"OverlyLongMethod", "OverlyNestedMethod"}) // will be broken up
    public void generateInto(Sample sample, Random random, int centerX, int centerY, int centerZ) {
        /*
        Implementation taken from Bukkit's github, here be dragons
         */
        float f = random.nextFloat() * 3.1415927F;
        double d0 = (double) ((float) (centerX + 8) + MathHelper.sin(f) * (float) maxVeinSize / 8.0F);
        double d1 = (double) ((float) (centerX + 8) - MathHelper.sin(f) * (float) maxVeinSize / 8.0F);
        double d2 = (double) ((float) (centerZ + 8) + MathHelper.cos(f) * (float) maxVeinSize / 8.0F);
        double d3 = (double) ((float) (centerZ + 8) - MathHelper.cos(f) * (float) maxVeinSize / 8.0F);
        double d4 = (double) (centerY + random.nextInt(3) - 2);
        double d5 = (double) (centerY + random.nextInt(3) - 2);

        for (int l = 0; l <= maxVeinSize; ++l) {
            double d6 = d0 + (d1 - d0) * (double) l / (double) maxVeinSize;
            double d7 = d4 + (d5 - d4) * (double) l / (double) maxVeinSize;
            double d8 = d2 + (d3 - d2) * (double) l / (double) maxVeinSize;
            double d9 = random.nextDouble() * (double) maxVeinSize / 16.0D;
            double d10 =
                    (double) (MathHelper.sin((float) l * 3.1415927F / (float) maxVeinSize) + 1.0F) * d9 + 1.0D;
            double d11 =
                    (double) (MathHelper.sin((float) l * 3.1415927F / (float) maxVeinSize) + 1.0F) * d9 + 1.0D;
            int i1 = MathHelper.floor(d6 - d10 / 2.0D);
            int j1 = MathHelper.floor(d7 - d11 / 2.0D);
            int k1 = MathHelper.floor(d8 - d10 / 2.0D);
            int l1 = MathHelper.floor(d6 + d10 / 2.0D);
            int i2 = MathHelper.floor(d7 + d11 / 2.0D);
            int j2 = MathHelper.floor(d8 + d10 / 2.0D);

            for (int k2 = i1; k2 <= l1; ++k2) {
                double d12 = ((double) k2 + 0.5D - d6) / (d10 / 2.0D);

                if (d12 * d12 < 1.0D) {
                    for (int l2 = j1; l2 <= i2; ++l2) {
                        double d13 = ((double) l2 + 0.5D - d7) / (d11 / 2.0D);

                        if (d12 * d12 + d13 * d13 < 1.0D) {
                            for (int i3 = k1; i3 <= j2; ++i3) {
                                double d14 = ((double) i3 + 0.5D - d8) / (d10 / 2.0D);

                                if (d12 * d12 + d13 * d13 + d14 * d14 < 1.0D && sample.hasBlock(k2, l2, i3)
                                        && sample.getBlock(k2, l2, i3).getType() == blockTypeToReplace) {
                                    sample.setBlock(k2, l2, i3, blockTypeToGenerate);
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

File: src/main/kotlin/org/hildan/minecraft/mining/optimizer/ore/WorldGenMinable.kt
package org.hildan.minecraft.mining.optimizer.ore

import org.hildan.minecraft.mining.optimizer.blocks.Sample

import java.util.Random

internal class WorldGenMinable(
    private val blockTypeToGenerate: BlockType
) {
    private val maxVeinSize: Int = blockTypeToGenerate.maxVeinSize
    private val maxVeinSizeDouble: Double = blockTypeToGenerate.maxVeinSize.toDouble()
    private val blockTypeToReplace: BlockType = BlockType.STONE

    /*
     Implementation retro-engineered from Bukkit's github, here be dragons.
     https://github.com/Bukkit/mc-dev/blob/master/net/minecraft/server/WorldGenMinable.java
     */
    fun generateInto(sample: Sample, random: Random, centerX: Int, centerY: Int, centerZ: Int) {

        // looks like random orientation
        val a = random.nextDouble() * Math.PI
        val sinA = MathHelper.sin(a)
        val cosA = MathHelper.cos(a)

        // looks like bounds calculation based on orientation
        val shiftedCenterX = (centerX + 8).toDouble()
        val shiftedCenterZ = (centerZ + 8).toDouble()
        val highX = (shiftedCenterX + sinA * maxVeinSizeDouble / 8.0)
        val lowX = (shiftedCenterX - sinA * maxVeinSizeDouble / 8.0)
        val highZ = (shiftedCenterZ + cosA * maxVeinSizeDouble / 8.0)
        val lowZ = (shiftedCenterZ - cosA * maxVeinSizeDouble / 8.0)
        val highY = (centerY + random.nextInt(3) - 2).toDouble()
        val lowY = (centerY + random.nextInt(3) - 2).toDouble()

        for (size in 0..maxVeinSize) {
            val sizeD = size.toDouble()

            fun scale(d: Double) = d * sizeD / maxVeinSizeDouble

            val scaledX = highX + scale(lowX - highX)
            val scaledY = highY + scale(lowY - highY)
            val scaledZ = highZ + scale(lowZ - highZ)

            val randSize = random.nextDouble() * maxVeinSizeDouble / 16.0
            val sinScaledPi = MathHelper.sin(scale(Math.PI))

            val diameter = (sinScaledPi + 1.0) * randSize + 1.0
            val radius = diameter / 2.0
            val minX = MathHelper.floor(scaledX - radius)
            val minY = MathHelper.floor(scaledY - radius)
            val minZ = MathHelper.floor(scaledZ - radius)
            val maxX = MathHelper.floor(scaledX + radius)
            val maxY = MathHelper.floor(scaledY + radius)
            val maxZ = MathHelper.floor(scaledZ + radius)

            for (x in minX..maxX) {
                val d12 = (x.toDouble() + 0.5 - scaledX) / radius
                val sqD12 = d12 * d12

                if (sqD12 < 1.0) {
                    for (y in minY..maxY) {
                        val d13 = (y.toDouble() + 0.5 - scaledY) / radius
                        val sqD13 = d13 * d13

                        if (sqD12 + sqD13 < 1.0) {
                            for (z in minZ..maxZ) {
                                val d14 = (z.toDouble() + 0.5 - scaledZ) / radius
                                val sqD14 = d14 * d14

                                if (sqD12 + sqD13 + sqD14 < 1.0 && sample.contains(x, y, z)
                                    && sample.getBlock(x, y, z).type == blockTypeToReplace
                                ) {
                                    sample.setBlock(x, y, z, blockTypeToGenerate)
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}
